### PR TITLE
[MLIR][tensor] Improve `tensor.pack` verifier to catch more cases with unconditional runtime errors

### PR DESCRIPTION
**Pull request by srcarroll**
Previously, the `tensor.pack` verifier detects unconditional runtime errors only when tile sizes are static.  Now, dynamic tiles are considered and we only require that the input and either corresponding tile or output size are static to determine if it will unconditionally produce errors at runtime.

### CODE REVIEW

**Comment by joker-eph** [comment]
> Improve tensor.pack verifier to catch more UB cases

I'm confused to see "verifier" and "UB" in the same sentence: these are non-overlapping concepts. If the verifier defines an IR to be invalid, it can't be UB. Only valid IR can be UB since UB is a runtime concept.

**Comment by srcarroll** [comment]
>  If the verifier defines an IR to be invalid, it can't be UB. Only valid IR can be UB since UB is a runtime concept.

Good point. Sorry I tend to abuse terminology sometimes.  Also the current code uses this phrasing. I'll try to come up with a better description, but any thoughts?

**Comment by nicolasvasilache** [comment]
I would just drop any notion of UB for the static case: it is invalid for an op to be a partial tile and to not provide a padding value.

In the dynamic case, we could always require a padding value (if feasible, I forget).

**Comment by joker-eph** [comment]
> I would just drop any notion of UB for the static case: it is invalid for an op to be a partial tile and to not provide a padding value.

This is fine, just to keep in mind that canonicalization/folder must include the same checks as the verifier when trying to constant fold the dynamic case into the static one.

**Comment by srcarroll** [comment]
> In the dynamic case, we could always require a padding value (if feasible, I forget

A padding value isn't required if the corresponding output dim is static and divides the the corresponding static input dim. That's the whole point of this PR. Unless I'm mistaken about something.

**Comment by chelini** [comment]
> I would just drop any notion of UB for the static case: it is invalid for an op to be a partial tile and to not provide a padding value.
> 
> In the dynamic case, we could always require a padding value (if feasible, I forget).

Would it make sense to keep the padding behavior consistent between the static and dynamic behavior, having padding optional in both cases?

**Comment by hanhanW** [comment]
> Would it make sense to keep the padding behavior consistent between the static and dynamic behavior, having padding optional in both cases?

+1 on keeping the padding behavior consistent between static and dynamic cases. It is valid to me. I don't have a specific use case, but I think user can allocate big enough buffer based on inner_tile_sizes.

**Comment by srcarroll** [comment]
> Would it make sense to keep the padding behavior consistent between the static and dynamic behavior, having padding optional in both cases?

I dont think I understand this. Can you elaborate?  I thought that's what I was doing.  That is, in current `main`, when tile sizes are static we require padding when tile size doesn't divide input size.  The proprosed changes here extend this to the dynamic tile case, but require padding when output size doesn't divide input size.  That's consistent by my definition so obviously I'm missing something here.

**Comment by chelini** [comment]
> > Would it make sense to keep the padding behavior consistent between the static and dynamic behavior, having padding optional in both cases?
> 
> I dont think I understand this. Can you elaborate? I thought that's what I was doing. That is, in current `main`, when tile sizes are static we require padding when tile size doesn't divide input size. The proprosed changes here extend this to the dynamic tile case, but require padding when output size doesn't divide input size. That's consistent by my definition so obviously I'm missing something here.

I answered Nicolas's comment and argued that keeping the padding optional in the dynamic case would be better (unless I misunderstood his message). I see op verification this way: If padding is specified, we will pad along high dimensions to make the tile complete; the tile will divide the dimension and the IR is valid. If padding is not specified, we have two options:
- If we have enough static information to prove that a tile does not divide perfectly a dimension, the verifier should emit an error. As far as I can see in this PR, you are going in this direction: input and output are constant, and you are erroring out because the tile cannot fully divide the dimension unless we pad.
- If we have dynamic behaviors and it is impossible to prove that the tile perfectly divides the dimension, the verifier should not emit any error, and it is an undefined behavior at runtime if the tile does not divide unless the padding attribute is specified.

**Comment by srcarroll** [comment]
@nicolasvasilache any thoughts?

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 1]
```diff
@@ -3746,14 +3746,27 @@ SmallVector<int64_t> PackOp::getStaticTiles() {
...
-    if (!constantTile)
-      continue;
-    if (inputShape[pos] % (*constantTile) != 0)
+
+    if (!constantTile) {
+      if (!ShapedType::isDynamic(outputTileSizes[pos]) &&
+          (inputShape[pos] % outputTileSizes[pos] != 0))
+        return true;
+    } else if (inputShape[pos] % (*constantTile) != 0)
       return true;
```
Comment (hanhanW) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:3770]: We should use braces for `else if`, which keeps it uniforms with `if` block.

https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

**Review by hanhanW — COMMENTED** [review]
This is very rare and I can't think a scenario about it... FYI that the changes about `requirePaddingValue` will break all the downstream projects that use the method. I'm fine with the change, but do we really have a need for this?

```
%0 = tensor.pack %input inner_dims_pos = [1, 0] inner_tiles = [%tile_size_0, %tile_size_1] into %output : tensor<256x128xf32> -> tensor<10x8x?x?xf32>
```

**Comment by nicolasvasilache** [comment]
My argument goes this way:
- we can remove UB by always requiring the padding value to be specified. It generally seems desirable to me to avoid needlessly introducing UB and special runtime cases.
- in the static case, when we have enough info, we can relax this to omit the padding value iff we can statically prove what we need to.

Ultimately, this smells like the "in_bounds" attribute that LLVM has on load/store and that we also added to vector.transfer_read/write.

The situation seems close enough to me to justify a similar attribute addition.

```
 I'm fine with the change, but do we really have a need for this?

%0 = tensor.pack %input inner_dims_pos = [1, 0] inner_tiles = [%tile_size_0, %tile_size_1] into %output : tensor<256x128xf32> -> tensor<10x8x?x?xf32>
```

Not planning to use it myself right now but I can see how a transformation guaranteeing that a dynamic tile size will not result in padding seems pretty useful. We would probably be better off with an "in_bounds" style attribute for this.

```
FYI that the changes about requirePaddingValue will break all the downstream projects that use the method. 
```

This has never been, and continues to not be, a factor for upstream MLIR. If it is the right thing to do we should, if not that's ok. Arguing design positions based on breaking downstream projects is dangerous: it gives the wrong incentives as it pigeonholes evolution.

**Comment by rengolin** [comment]
> This has never been, and continues to not be, a factor for upstream MLIR. If it is the right thing to do we should, if not that's ok. Arguing design positions based on breaking downstream projects is dangerous: it gives the wrong incentives as it pigeonholes evolution.

Strong support for this statement.

> * we can remove UB by always requiring the padding value to be specified. It generally seems desirable to me to avoid needlessly introducing UB and special runtime cases.

I wonder what is the value of requiring an actual value at run time. IIUC, the padding is just the remainder of the division of the run time dimension by the run time tiling factor, on the last tile. Anything else would make no sense. Requiring users/lowering to add a `mod` op for the padding value seems redundant. Just a flag to pad to the nearest multiple on the last tile would do, probably.

If the source language "defines" undefined behaviour for run time cases (like C++), then we let it by pass. If not, then they're responsible for lowering run time checks to IR and do what they normally would do (crash, trap, raise exception). It is not a job for the IR to know what.

**Comment by srcarroll** [comment]
> I'm fine with the change, but do we really have a need for this?
@hanhanW To be honest, I personally don't need it. this was just a follow up to the discussion in this PR https://github.com/llvm/llvm-project/pull/76003 . Which I also don't personally need. 

> we can remove UB by always requiring the padding value to be specified. It generally seems desirable to me to avoid needlessly introducing UB and special runtime cases.
seems reasonable to me. by the time i started touching this code there was already a precedence to detect UB, so I was just trying to finish that off

> in the static case, when we have enough info, we can relax this to omit the padding value iff we can statically prove what we need to.

if i'm not mistaken, the proposed changes should cover all the cases

> We would probably be better off with an "in_bounds" style attribute for this.
I'm not opposed to that, but honestly don't know how that would work yet.

> I wonder what is the value of requiring an actual value at run time. IIUC, the padding is just the remainder of the division of the run time dimension by the run time tiling factor, on the last tile. Anything else would make no sense. Requiring users/lowering to add a mod op for the padding value seems redundant. Just a flag to pad to the nearest multiple on the last tile would do, probably.

you are referring to the padding size, not the padding value (the value filling the padded regions). A padding value is always required when padding is needed. and padding is needed when the the tile size doesn't divide the input.

**Review by hanhanW — APPROVED** [review]
LGTM, thanks for the patch. Now we should cover all the cases in the verifier.

> Ultimately, this smells like the "in_bounds" attribute that LLVM has on load/store and that we also added to vector.transfer_read/write.

IMO. we don't want to propagate "in_bounds" attribute to high level operations. This looks like a good stop point to me at tensor level, and we handle the case in vectorization.

**Pull request merged**